src/main/java/com/softserve/repository/SortOrderRepository.java

```diff
@@ -0,0 +1,113 @@
+package com.softserve.repository;
+
+import com.softserve.entity.interfaces.SortableOrder;
+import com.softserve.exception.EntityNotFoundException;
+import com.softserve.exception.SortOrderNotExistsException;
+import lombok.extern.slf4j.Slf4j;
+import org.hibernate.SessionFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Repository;
+
+import javax.persistence.TypedQuery;
+import java.util.Optional;
+
+@Slf4j
+@Repository
+@Scope(value = ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class SortOrderRepository<T extends SortableOrder> {
+    public SortOrderRepository() {
+    }
+
+    private Class<?> tClass = SortOrderRepository.class;
+    private SessionFactory sessionFactory;
+
+    @Autowired
+    public SortOrderRepository(SessionFactory sessionFactory) {
+        this.sessionFactory = sessionFactory;
+    }
+
+    public T createAfterOrder(T t, Long afterId) {
+        log.info("Entered createAfterOrder({}, {})", afterId, t);
+        Integer maxOrder = getMaxSortOrder().orElse(0);
+        Integer order;
+        if (afterId != null && afterId != 0) {
+            order = getSortOrderById(afterId).orElseThrow(() -> new SortOrderNotExistsException(tClass, afterId)) + 1;
+            t.setSortOrder(order);
+            changeOrderOffset(order, maxOrder + 1);
+        } else {
+            t.setSortOrder(1);
+            changeOrderOffset(0, maxOrder + 1);
+        }
+        sessionFactory.getCurrentSession().save(t);
+        return t;
+    }
+
+    public T updateAfterOrder(T t, Long afterId) {
+        log.info("Entered updateAfterOrder({}, {})", t, afterId);
+        if (!isExistsById(t.getId())) {
+            throw new EntityNotFoundException(tClass);
+        }
+        if (t.getId().equals(afterId)) {
+            return t;
+        }
+        Integer maxOrder = getMaxSortOrder().orElse(0);
+        if (afterId != null && afterId != 0) {
+            Integer lowerBound = getSortOrderById(afterId).orElseThrow(() -> new SortOrderNotExistsException(tClass, afterId)) + 1;
+            Integer upperBound = Optional.ofNullable(t.getSortOrder()).orElse(maxOrder + 1) + 1;
+            t.setSortOrder(lowerBound);
+            changeOrderOffset(lowerBound, upperBound);
+        } else {
+            t.setSortOrder(1);
+            changeOrderOffset(0, maxOrder + 1);
+        }
+        sessionFactory.getCurrentSession().update(t);
+        return t;
+    }
+
+    public boolean isExistsById(Long id) {
+        log.info("In isExistsById(id = [{}])", id);
+        return (boolean) sessionFactory.getCurrentSession()
+                .createQuery(
+                        "SELECT (count(*) > 0)"
```

Comment:
SELECT 1 or SELECT EXISTS could be used, I'd recommend the first approach, in that case uniqueResult() != null can be checked